elfloader-tool/src/arch-riscv/crt0.S

```diff
@@ -33,6 +33,11 @@ _start:
 /* a0 should have hart id, store it in temp so as not to clobber from HSM calls */
   mv t3, a0
 
+#ifdef CONFIG_IMAGE_BINARY
+/* Clear the BSS before we get to do anything more specific */
```

Comment:
* I consider this comment a bit too vague. TheC compiler expects bss to be filled with zeros, but for any low level hand written asm code this is not necessarily true.
* Looking at the clear_bss() implementation in C, that is just a native zero-byte-by-byte loop it might be more elegant to inline this code here and then have full control over the register usage. Then we can preserved registers easily and don't have to worry about a calling convention.
* There is a core register for the hard ID, so why to we actual care about preserving this values? We could simply read this registers when needed. Otherwise the code need some explanation why this can't be used and we need to preserve a value passed by another boot stage.

Reply:
> I consider this comment a bit too vague. TheC compiler expects bss to be filled with zeros, but for any low level hand written asm code this is not necessarily true. Furthermore, you can write (early boot) C code that does not does not use the bss, so clearing bss is not necessarily needed if the early boot code is clearly defined.

The early C boot code in SMP requires the use of a variable in BSS which is why we clear the BSS very early on. We can optimise this so that only the main HART runs it, but synchronisation with the other cores is a little difficult.

> Looking at the clear_bss() implementation in C, that is just a native zero-byte-by-byte loop it might be more elegant to inline this code here and then have full control over the register usage. Then we can preserved registers easily and don't have to worry about a calling convention.

Er, sure. I just wasn't comfortable writing RISC-V assembly and there just happened to be a `clear_bss()` shared C implementation.

> There is a core register for the hard ID, so why to we actual care about preserving this values? We could simply read this registers when needed. Otherwise the code need some explanation why this can't be used and we need to preserve a value passed by another boot stage.

I'm not sure what this is referring to. Is this directed to the other commit?

> We don't have to zero the bss on all core and should not do this. Only the primary core should do this and the asm should contain the code to sort this out.

Re above to the comments above about optimisation and assembly.